mountpoint-s3/src/inode.rs

```diff
@@ -351,6 +351,59 @@ impl Superblock {
 
         Ok(LookedUp { inode, stat })
     }
+
+    /// Create a new directory inode
+    pub async fn createdir<OC: ObjectClient>(
+        &self,
+        client: &OC,
+        dir: InodeNo,
+        name: &OsStr,
+    ) -> Result<LookedUp, InodeError> {
+        trace!(parent=?dir, ?name, "createdir");
+
+        let existing = self.lookup(client, dir, name).await;
+        match existing {
+            Ok(lookup) => return Err(InodeError::FileAlreadyExists(lookup.inode.ino())),
+            Err(InodeError::FileDoesNotExist) => (),
+            Err(e) => return Err(e),
+        }
+
+        // Should be impossible to fail since [lookup] does this check, but let's be sure
+        let name = name
+            .to_str()
+            .ok_or_else(|| InodeError::InvalidFileName(name.to_owned()))?;
+
+        let parent_inode = self.inner.get(dir)?;
+        let mut parent_state = parent_inode.inner.sync.write().unwrap();
+
+        // Check again for the child now that the parent is locked, since we might have lost to a
+        // racing lookup. (It would be nice to lock the parent and *then* lookup, but we'd have to
+        // hold that lock across the remote API calls).
+        let InodeKindData::Directory { children, .. } = &mut parent_state.kind_data else {
+            return Err(InodeError::NotADirectory(dir));
+        };
+        if let Some(inode) = children.get(name) {
+            return Err(InodeError::FileAlreadyExists(inode.ino()));
+        }
+
+        let expiry = Instant::now(); // TODO local inode stats never expire?
+        let stat = InodeStat::for_directory(OffsetDateTime::now_utc(), expiry);
+        let kind = InodeKind::Directory;
+        let state = InodeState {
+            stat: stat.clone(),
+            kind_data: InodeKindData::Directory {
+                children: Default::default(),
+                writing_children: Default::default(),
+            },
+            write_status: WriteStatus::LocalUnopened,
```

Comment:
Not sure what status we should be using, but we might want to update this to `WriteStatus::Remote` when the first file is created.

Reply:
Transition from local to remote implemented in `finish_writing` of nested files.